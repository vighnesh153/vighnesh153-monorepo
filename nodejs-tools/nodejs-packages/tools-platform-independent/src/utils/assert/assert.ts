import { not } from '../not';

export function assert(condition: boolean, message = 'Assertion failed') {
  if (not(condition)) {
    throw new Error(message);
  }
}
