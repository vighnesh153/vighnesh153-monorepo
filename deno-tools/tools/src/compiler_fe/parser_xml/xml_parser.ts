import { not } from "@vighnesh153/tools";
import { assert } from "@std/assert";
import { nextToken, XmlTokenType } from "@/compiler_fe/lexer_xml/mod.ts";
import { ParserError } from "./parser_error.ts";
import {
  XmlCommentNode,
  XmlElementAttribute,
  type XmlExpression,
  XmlProgram,
  XmlPrologNode,
  XmlTagNode,
  XmlTextNode,
} from "./ast.ts";
import type { Lexer, Token } from "@/compiler_fe/lexer_core/mod.ts";

export class XmlParser {
  readonly #errors: Array<ParserError> = [];

  #currentToken!: Readonly<Token<XmlTokenType>>;
  #peekToken!: Readonly<Token<XmlTokenType>>;

  get errors(): Readonly<Array<ParserError>> {
    return this.#errors.map((error) => error.copy());
  }

  constructor(readonly lexer: Lexer<XmlTokenType>) {
    this.nextToken();
    this.nextToken();
  }

  parseProgram(): XmlProgram {
    const program = new XmlProgram();

    while (not(this.isCurrentToken(XmlTokenType.Eof))) {
      const statement = this.parseStatement();
      if (statement === null) {
        break;
      }
      program.addStatement(statement);
      this.nextToken();
    }

    return program;
  }

  addError(error: ParserError): void {
    this.#errors.push(error);
  }

  isCurrentToken(tokenType: XmlTokenType): boolean {
    return this.#currentToken.tokenType === tokenType;
  }

  isPeekToken(tokenType: XmlTokenType): boolean {
    return this.#peekToken.tokenType === tokenType;
  }

  expectPeek(tokenType: XmlTokenType): boolean {
    if (this.isPeekToken(tokenType)) {
      this.nextToken();
      return true;
    }
    this.addError(
      new ParserError({
        errorType: "UNEXPECTED_TOKEN",
        culpritToken: this.#peekToken,
      }),
    );
    return false;
  }

  private nextToken() {
    if (this.#peekToken) {
      this.#currentToken = this.#peekToken;
    }
    this.#peekToken = nextToken(this.lexer);
  }

  private parseStatement(): XmlExpression | null {
    if (this.isCurrentToken(XmlTokenType.LeftAngleBracket)) {
      if (this.isPeekToken(XmlTokenType.QuestionMark)) {
        return this.parseXmlPrologNode();
      }
      if (this.isPeekToken(XmlTokenType.Identifier)) {
        return this.parseXmlTagNode();
      }
      this.addError(
        new ParserError({
          errorType: "UNEXPECTED_TOKEN",
          culpritToken: this.#peekToken,
        }),
      );
      return null;
    }
    if (this.isCurrentToken(XmlTokenType.CommentLiteral)) {
      return this.parseXmlCommentNode();
    }
    if (this.isCurrentToken(XmlTokenType.TextNode)) {
      return this.parseXmlTextNode();
    }
    this.addError(
      new ParserError({
        errorType: "UNEXPECTED_TOKEN",
        culpritToken: this.#currentToken,
      }),
    );
    return null;
  }

  private parseXmlPrologNode(): XmlPrologNode | null {
    assert(
      this.isCurrentToken(XmlTokenType.LeftAngleBracket),
      `Shouldn't call parseXmlPrologNode when current token is not '<'`,
    );

    // move past "<"
    this.nextToken();

    assert(
      this.isCurrentToken(XmlTokenType.QuestionMark),
      `Shouldn't call parseXmlPrologNode when expression doesn't start with '<?'`,
    );

    if (not(this.expectPeek(XmlTokenType.Identifier))) {
      return null;
    }

    if (this.#currentToken.tokenLiteral !== "xml") {
      this.addError(
        new ParserError({
          errorType: "UNEXPECTED_PROLOG_TAG",
          culpritToken: this.#currentToken,
        }),
      );
      return null;
    }

    this.nextToken();

    const xmlPrologNode = new XmlPrologNode();

    while (true) {
      if (this.isCurrentToken(XmlTokenType.Eof)) {
        this.addError(
          new ParserError({
            culpritToken: this.#currentToken,
            errorType: "UNEXPECTED_EOF",
          }),
        );
        return null;
      }

      if (this.isCurrentToken(XmlTokenType.QuestionMark)) {
        if (not(this.expectPeek(XmlTokenType.RightAngleBracket))) {
          return null;
        }
        return xmlPrologNode;
      }

      const attribute = this.parseAttribute();
      if (attribute === null) {
        return null;
      }
      xmlPrologNode.addAttribute(attribute);
      this.nextToken();
    }
  }

  private parseXmlTagNode(): XmlTagNode | null {
    assert(
      this.isCurrentToken(XmlTokenType.LeftAngleBracket),
      `Shouldn't call parseXmlTagNode when current token is not '<'`,
    );

    this.nextToken();

    assert(
      this.isCurrentToken(XmlTokenType.Identifier),
      `Shouldn't call parseXmlTagNode when statement doesn't start with "<IDENTIFIER"`,
    );

    const xmlTagNode = new XmlTagNode();
    xmlTagNode.addNamespace(this.#currentToken);
    this.nextToken();

    while (this.isCurrentToken(XmlTokenType.Colon)) {
      if (!this.expectPeek(XmlTokenType.Identifier)) {
        return null;
      }
      xmlTagNode.addNamespace(this.#currentToken);
      this.nextToken();
    }

    while (true) {
      if (this.isCurrentToken(XmlTokenType.Eof)) {
        this.addError(
          new ParserError({
            culpritToken: this.#currentToken,
            errorType: "UNEXPECTED_EOF",
          }),
        );
        return null;
      }

      if (this.isCurrentToken(XmlTokenType.ForwardSlash)) {
        break;
      }

      if (this.isCurrentToken(XmlTokenType.RightAngleBracket)) {
        break;
      }

      const attribute = this.parseAttribute();
      if (attribute === null) {
        return null;
      }
      xmlTagNode.addAttribute(attribute);
      this.nextToken();
    }

    if (this.isCurrentToken(XmlTokenType.ForwardSlash)) {
      if (not(this.expectPeek(XmlTokenType.RightAngleBracket))) {
        return null;
      }
      return xmlTagNode;
    }

    assert(
      this.isCurrentToken(XmlTokenType.RightAngleBracket),
      `Expected ">" found ${this.#currentToken.tokenLiteral}`,
    );

    // move past ">"
    this.nextToken();

    // parse children
    while (true) {
      if (this.isPeekToken(XmlTokenType.Eof)) {
        this.addError(
          new ParserError({
            culpritToken: this.#peekToken,
            errorType: "UNEXPECTED_EOF",
          }),
        );
        return null;
      }

      if (
        this.isCurrentToken(XmlTokenType.LeftAngleBracket) &&
        this.isPeekToken(XmlTokenType.ForwardSlash)
      ) {
        break;
      }

      const statement = this.parseStatement();
      if (statement === null) {
        return null;
      }
      xmlTagNode.addChild(statement);
      this.nextToken();
    }

    assert(
      this.isCurrentToken(XmlTokenType.LeftAngleBracket),
      `Expected "<" found ${this.#currentToken.tokenLiteral}`,
    );

    // Move past "<"
    this.nextToken();

    if (not(this.expectPeek(XmlTokenType.Identifier))) {
      return null;
    }

    const closingTagNamespaces = [this.#currentToken];
    this.nextToken();

    while (this.isCurrentToken(XmlTokenType.Colon)) {
      if (!this.expectPeek(XmlTokenType.Identifier)) {
        return null;
      }
      closingTagNamespaces.push(this.#currentToken);
      this.nextToken();
    }

    const openingTagName = xmlTagNode.namespaces.map((part) =>
      part.tokenLiteral
    ).join(":");
    const closingTagName = closingTagNamespaces.map((ns) => ns.tokenLiteral)
      .join(":");
    if (openingTagName !== closingTagName) {
      this.addError(
        new ParserError({
          culpritToken: {
            ...closingTagNamespaces[0],
            tokenLiteral: closingTagName,
          },
          errorType: "UNEXPECTED_CLOSING_TAG_LITERAL",
        }),
      );
      return null;
    }

    if (not(this.isCurrentToken(XmlTokenType.RightAngleBracket))) {
      this.addError(
        new ParserError({
          culpritToken: this.#currentToken,
          errorType: "UNEXPECTED_TOKEN",
        }),
      );
      return null;
    }

    return xmlTagNode;
  }

  private parseAttribute(): XmlElementAttribute | null {
    if (not(this.isCurrentToken(XmlTokenType.Identifier))) {
      this.addError(
        new ParserError({
          culpritToken: this.#currentToken,
          errorType: "UNEXPECTED_TOKEN",
        }),
      );
      return null;
    }

    const namespaces: Readonly<Token<XmlTokenType>>[] = [this.#currentToken];

    while (true) {
      if (this.isPeekToken(XmlTokenType.Eof)) {
        this.addError(
          new ParserError({
            culpritToken: this.#peekToken,
            errorType: "UNEXPECTED_EOF",
          }),
        );
        return null;
      }
      if (this.isPeekToken(XmlTokenType.Equals)) {
        break;
      }

      if (not(this.expectPeek(XmlTokenType.Colon))) {
        return null;
      }
      if (not(this.expectPeek(XmlTokenType.Identifier))) {
        return null;
      }
      namespaces.push(this.#currentToken);
    }

    assert(
      this.isPeekToken(XmlTokenType.Equals),
      `Expected "=" found ${this.#peekToken.tokenLiteral}`,
    );

    // we know next token is equals. move to that.
    this.nextToken();

    if (not(this.expectPeek(XmlTokenType.StringLiteral))) {
      return null;
    }

    return new XmlElementAttribute(namespaces, this.#currentToken);
  }

  private parseXmlCommentNode(): XmlCommentNode {
    assert(
      this.isCurrentToken(XmlTokenType.CommentLiteral),
      `Shouldn't call parseXmlCommentNode when current token is not a comment token`,
    );

    return new XmlCommentNode(this.#currentToken);
  }

  private parseXmlTextNode(): XmlTextNode {
    assert(
      this.isCurrentToken(XmlTokenType.TextNode),
      `Shouldn't call parseXmlTextNode when current token is not a text token`,
    );

    return new XmlTextNode(this.#currentToken);
  }
}
