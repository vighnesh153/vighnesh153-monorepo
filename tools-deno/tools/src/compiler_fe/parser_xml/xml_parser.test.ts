import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import {
  Lexer,
  LexerInputReader,
  StringLexerInput,
} from "@/compiler_fe/lexer_core/mod.ts";
import { XmlTokenType } from "@/compiler_fe/lexer_xml/mod.ts";
import { XmlParser } from "./xml_parser.ts";
import type { XmlProgram } from "./ast.ts";

function parseProgram(input: string): [XmlParser, XmlProgram] {
  const stringInput = new StringLexerInput(input);
  const inputReader = new LexerInputReader(stringInput);
  const lexer = new Lexer<XmlTokenType>(inputReader);
  const parser = new XmlParser(lexer);

  return [parser, parser.parseProgram()];
}

Deno.test("XmlParser should parse empty xml input", () => {
  const [parser, program] = parseProgram("");

  assertEquals(parser.errors, []);
  assertEquals(program.statements, []);
});

Deno.test("XmlParser should parse prolog tag without attributes", async (t) => {
  const [parser, program] = parseProgram("< ? xml ? >");

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse prolog tag with attributes", async (t) => {
  const [parser, program] = parseProgram(
    '< ? xml version  =  "1.0" encoding ="UTF-8" ? >',
  );

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse empty self-closing tag", async (t) => {
  const [parser, program] = parseProgram("< manifest / >");

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse empty tag", async (t) => {
  const [parser, program] = parseProgram("< manifest > < /   manifest>");

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse tag with single simple attribute", async (t) => {
  const [parser, program] = parseProgram(
    '< manifest package="com.pokemon.pikachu" / >',
  );

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse tag with multiple simple attributes", async (t) => {
  const [parser, program] = parseProgram(
    '< manifest package= "com.pokemon.pikachu" attr2  =   "2.0" / >',
  );

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse tag with colon separated attributes", async (t) => {
  const [parser, program] = parseProgram(
    `<uses-feature      android:name   =    
    
    "android.hardware.touchscreen" 
    
    android:required      =
    "false" />`,
  );

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test(
  "XmlParser should be able to read tags with namespaces and classes",
  async (t) => {
    const [parser, program] = parseProgram(
      "< namespace1:namespace2.class1.class2:namespace3.class3.class4 > </ namespace1:namespace2.class1.class2:namespace3.class3.class4>",
    );

    assertEquals(parser.errors.length, 0);
    assertEquals(program.statements.length, 1);
    await assertSnapshot(t, program.toString(0));
  },
);

Deno.test("XmlParser should parse naked text node", async (t) => {
  const [parser, program] = parseProgram("pikachu");

  assertEquals(parser.errors.length, 0);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse tag with children", async (t) => {
  const [parser, program] = parseProgram(
    `
    <?xml 
    
    
    
    
    version="1.0" encoding="utf-8" ?>
    <manifest xmlns:android="http://schemas.android.com/apk/res/android"
        package="com.pokemon.charizard">
        <uses-feature android:name="android.hardware.touchscreen" android:required="false" />
        <!-- 
        
        
        
        <uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" /> 
        
        
        -->

        <application
            android:allowBackup="true"
            android:label="Pikachu"
            android:supportsRtl="true"
                        >
            <receiver
                android:name=".MyReceiver"
                android:enabled="true"
                android:exported="true">
                <intent-filter>
                    <action android:name="SOME_INTENT_FILTER_1" />
                    <action android:name="SOME_INTENT_FILTER_2" 
                    />
                </intent-filter>
            </receiver>

            <activity
                android:name=".MainActivity"
                android:exported="true"
                android:launchMode="singleTask">
                <intent-filter>
                    <action android:name="android.intent.action.MAIN" />
                    <category android:name="android.intent.category.LAUNCHER" />
                </intent-filter>
                <intent-filter>
                    <action android:name="android.intent.action.MAIN" />
                    <category android:name="android.intent.category.LEANBACK_LAUNCHER" />
                </intent-filter>
            </activity>
      </application>
    </manifest>

    `,
  );

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 2);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should parse tag with text node children and attributes", async (t) => {
  const [parser, program] = parseProgram(`
    <    pokemon >
      <name
    >
      Pikachu<
    /
    
    name
>
      <
types    comma-separated   =  "true"    >    
      electric, god<  /
      types>
      <
      desc
      >
        
  Best pokemon ever!!!
      
                      </
                  desc    ><
                      /pokemon
                      >
    `);

  assertEquals(parser.errors, []);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(t, program.toString(0));
});

Deno.test("XmlParser should return error if incorrect token position", () => {
  const [parser, program] = parseProgram("?");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 1,
      lineNumber: 1,
      tokenLiteral: "?",
      tokenType: XmlTokenType.QuestionMark,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if EOF while reading attribute", () => {
  const [parser, program] = parseProgram("< manifest fruit");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 16,
      lineNumber: 1,
      tokenLiteral: "Eof",
      tokenType: XmlTokenType.Eof,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if identifier token after identifier token instead of colon while parsing attribute", () => {
  const [parser, program] = parseProgram(
    '< manifest pokemon fruit="orange" / >',
  );

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 20,
      lineNumber: 1,
      tokenLiteral: "fruit",
      tokenType: XmlTokenType.Identifier,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if equals token after colon token instead of identifier while parsing attribute", () => {
  const [parser, program] = parseProgram('< manifest fruit:="orange" / >');

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 17,
      lineNumber: 1,
      tokenLiteral: ":",
      tokenType: XmlTokenType.Illegal,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if identifier token after equals while parsing attribute", () => {
  const [parser, program] = parseProgram("< manifest fruit= orange / >");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 19,
      lineNumber: 1,
      tokenLiteral: "orange",
      tokenType: XmlTokenType.Identifier,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if identifier token after equals while parsing attribute in xml prolog", () => {
  const [parser, program] = parseProgram("<? xml fruit= orange ?>");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 15,
      lineNumber: 1,
      tokenLiteral: "orange",
      tokenType: XmlTokenType.Identifier,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if EOF while parsing xml prolog", () => {
  const [parser, program] = parseProgram("<? xml  ");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 8,
      lineNumber: 1,
      tokenLiteral: "Eof",
      tokenType: XmlTokenType.Eof,
    },
    errorType: "UNEXPECTED_EOF",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if EOF while parsing xml tag", () => {
  const [parser, program] = parseProgram("< manifest  ");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 12,
      lineNumber: 1,
      tokenLiteral: "Eof",
      tokenType: XmlTokenType.Eof,
    },
    errorType: "UNEXPECTED_EOF",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if incorrect token after closing question mark in xml prolog", () => {
  const [parser, program] = parseProgram("<? xml  ? <");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 11,
      lineNumber: 1,
      tokenLiteral: "<",
      tokenType: XmlTokenType.LeftAngleBracket,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error while parsing xml prolog if incorrect token after <?", () => {
  const [parser, program] = parseProgram("<? ?>");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 4,
      lineNumber: 1,
      tokenLiteral: "?",
      tokenType: XmlTokenType.QuestionMark,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error while parsing xml prolog if tag is not xml", () => {
  const [parser, program] = parseProgram("<? manifest  ?>");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 4,
      lineNumber: 1,
      tokenLiteral: "manifest",
      tokenType: XmlTokenType.Identifier,
    },
    errorType: "UNEXPECTED_PROLOG_TAG",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if incorrect token while parsing self closing xml tag", () => {
  const [parser, program] = parseProgram("< manifest  / <");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 15,
      lineNumber: 1,
      tokenLiteral: "<",
      tokenType: XmlTokenType.LeftAngleBracket,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if EOF while parsing children of xml tag", () => {
  const [parser, program] = parseProgram("< manifest >");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 12,
      lineNumber: 1,
      tokenLiteral: "Eof",
      tokenType: XmlTokenType.Eof,
    },
    errorType: "UNEXPECTED_EOF",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if error while parsing children of xml tag", () => {
  const [parser, program] = parseProgram(
    "< manifest > <child-tag /< </manifest>",
  );

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 26,
      lineNumber: 1,
      tokenLiteral: "<",
      tokenType: XmlTokenType.LeftAngleBracket,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return error if incorrect token at end of tag", () => {
  const [parser, program] = parseProgram("< manifest > </ ?");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 17,
      lineNumber: 1,
      tokenLiteral: "?",
      tokenType: XmlTokenType.QuestionMark,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test(`XmlParser should return error if start and close tag don't match`, () => {
  const [parser, program] = parseProgram("< manifest > </ pokemon>");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 17,
      lineNumber: 1,
      tokenLiteral: "pokemon",
      tokenType: XmlTokenType.Identifier,
    },
    errorType: "UNEXPECTED_CLOSING_TAG_LITERAL",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return errors if equals without identifier", () => {
  const [parser, program] = parseProgram(
    '< manifest ="com.pokemon.pikachu" / >',
  );

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 12,
      lineNumber: 1,
      tokenLiteral: "=",
      tokenType: XmlTokenType.Equals,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test('XmlParser should return errors if "<" instead of ">" in closing tag literal', () => {
  const [parser, program] = parseProgram("< tag > </ tag <");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 16,
      lineNumber: 1,
      tokenLiteral: "<",
      tokenType: XmlTokenType.LeftAngleBracket,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test("XmlParser should return errors if nameless tag", () => {
  const [parser, program] = parseProgram("< >");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 3,
      lineNumber: 1,
      tokenLiteral: ">",
      tokenType: XmlTokenType.RightAngleBracket,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test('XmlParser should return error when unexpected token after ":" in tag name', () => {
  const [parser, program] = parseProgram("< ns1:ns2:  /> ");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 10,
      lineNumber: 1,
      tokenLiteral: ":",
      tokenType: XmlTokenType.Illegal,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test(`XmlParser should return error if start and end tag names don't match`, () => {
  const [parser, program] = parseProgram("< ns1:ns2  > </ ns3:ns4> ");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 17,
      lineNumber: 1,
      tokenLiteral: "ns3:ns4",
      tokenType: XmlTokenType.Identifier,
    },
    errorType: "UNEXPECTED_CLOSING_TAG_LITERAL",
  });
  assertEquals(program.statements.length, 0);
});

Deno.test('XmlParser should return error when unexpected token after ":" in in closing tag name', () => {
  const [parser, program] = parseProgram("< ns1:ns2  > </ ns1:ns2: >");

  assertEquals(parser.errors.length, 1);
  assertEquals(parser.errors[0].serialized(), {
    culpritToken: {
      columnNumber: 24,
      lineNumber: 1,
      tokenLiteral: ":",
      tokenType: XmlTokenType.Illegal,
    },
    errorType: "UNEXPECTED_TOKEN",
  });
  assertEquals(program.statements.length, 0);
});
