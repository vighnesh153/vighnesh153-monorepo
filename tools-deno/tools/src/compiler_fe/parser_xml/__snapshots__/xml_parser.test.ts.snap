export const snapshot = {};

snapshot[`XmlParser should parse prolog tag without attributes 1`] = `"<?xml?>"`;

snapshot[`XmlParser should parse prolog tag with attributes 1`] = `'<?xml version="1.0" encoding="UTF-8"?>'`;

snapshot[`XmlParser should parse empty self-closing tag 1`] = `"<manifest />"`;

snapshot[`XmlParser should parse empty tag 1`] = `"<manifest />"`;

snapshot[`XmlParser should parse tag with single simple attribute 1`] = `'<manifest package="com.pokemon.pikachu" />'`;

snapshot[`XmlParser should parse tag with multiple simple attributes 1`] = `
'<manifest
    package="com.pokemon.pikachu"
    attr2="2.0" />'
`;

snapshot[`XmlParser should parse tag with colon separated attributes 1`] = `
'<uses-feature
    android:name="android.hardware.touchscreen"
    android:required="false" />'
`;

snapshot[`XmlParser should be able to read tags with namespaces and classes 1`] = `"<namespace1:namespace2.class1.class2:namespace3.class3.class4 />"`;

snapshot[`XmlParser should parse naked text node 1`] = `"pikachu"`;

snapshot[`XmlParser should parse tag with children 1`] = `
'<?xml version="1.0" encoding="utf-8"?>
<manifest
    xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.pokemon.charizard">
    <uses-feature
        android:name="android.hardware.touchscreen"
        android:required="false" />
    <!-- <uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" /> -->
    <application
        android:allowBackup="true"
        android:label="Pikachu"
        android:supportsRtl="true">
        <receiver
            android:name=".MyReceiver"
            android:enabled="true"
            android:exported="true">
            <intent-filter>
                <action android:name="SOME_INTENT_FILTER_1" />
                <action android:name="SOME_INTENT_FILTER_2" />
            </intent-filter>
        </receiver>
        <activity
            android:name=".MainActivity"
            android:exported="true"
            android:launchMode="singleTask">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LEANBACK_LAUNCHER" />
            </intent-filter>
        </activity>
    </application>
</manifest>'
`;

snapshot[`XmlParser should parse tag with text node children and attributes 1`] = `
'<pokemon>
    <name>Pikachu</name>
    <types comma-separated="true">electric, god</types>
    <desc>Best pokemon ever!!!</desc>
</pokemon>'
`;
